src/desktop.c

```diff
@@ -139,6 +139,57 @@ desktop_cycle_view(struct server *server, struct view *start_view,
 	return iter(&server->views, start_view, criteria);
 }
 
+/*
+ *  is either box overlapping the other by 80% of its own size?
+ */
+int
+is_mostly_overlapping(struct wlr_box *a, struct wlr_box *b)
+{
+	inline int min(int a, int b) { return (a < b) ? a : b; }
+	inline int max(int a, int b) { return (a > b) ? a : b; }
+
+	// calculate the overlap to determine if it's mostly within the region
+	int x_overlap = max(0, min(a->x + a->width, b->x + b->width)
+				- max(a->x, b->x));
+	int y_overlap = max(0, min(a->y + a->height, b->y + b->height)
+				- max(a->y, b->y));
+	int overlap_area = x_overlap * y_overlap;
+	int areaA = a->width * a->height;
+	int areaB = b->width * b->height;
+	return overlap_area > 0.80 * areaA || overlap_area > 0.80 * areaB;
+}
+
+struct view *
+desktop_cycle_view_in_region(struct server *server, struct view *active_view,
+		struct region *region)
+{
+	assert(region);
+
+	struct view *(*iter)(struct wl_list *head, struct view *view,
+		enum lab_view_criteria criteria);
+	bool forwards = true;
+
+	// if the currently active/focused window is within the region,
+	// then we want to cycle through in reverse order so that we
+	// cycle through all windows, and not just the two most recent.
+	if (active_view && is_mostly_overlapping(&active_view->current, &region->geo)) {
+		forwards = false;
```

Comment:
Rather than changing direction couldn't we just `cur = iter(&server->views, NULL, criteria);`?

(probably after initializing `iter` and `criteria`.

Reply:
I tried a few different ways here, and this one felt the best to me (and was simpler than my first attempt), with the window ordering for same-region focus cycling now feeling correct.  Give it another shot in a second.